File: src/test/java/nl/knaw/huygens/alexandria/creole/UtilitiesTest.java
package nl.knaw.huygens.alexandria.creole;

/*
 * #%L
 * alexandria-markup
 * =======
 * Copyright (C) 2016 - 2017 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import org.junit.Test;

import static nl.knaw.huygens.alexandria.AlexandriaAssertions.assertThat;
import static nl.knaw.huygens.alexandria.creole.Basics.id;
import static nl.knaw.huygens.alexandria.creole.Constructors.*;


public class UtilitiesTest extends CreoleTest {

  @Test
  public void testEmptyIsNullable() {
    Pattern p = Patterns.EMPTY;
    assertThat(p).isNullable();
  }

  @Test
  public void testNotAllowedIsNotNullable() {
    Pattern p = Patterns.NOT_ALLOWED;
    assertThat(p).isNotNullable();
  }

  @Test
  public void testTextIsNullable() {
    Pattern p = Patterns.TEXT;
    assertThat(p).isNullable();
  }

  @Test
  public void testChoiceNullability1() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Choice(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testChoiceNullability2() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Choice(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testChoiceNullability3() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Choice(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testChoiceNullability4() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Choice(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testInterleaveNullability1() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Interleave(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testInterleaveNullability2() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Interleave(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testInterleaveNullability3() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Interleave(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testInterleaveNullability4() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Interleave(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testConcurNullability1() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Concur(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testConcurNullability2() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Concur(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testConcurNullability3() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Concur(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testConcurNullability4() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Concur(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testPartitionNullability1() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p = new Patterns.Partition(p1);
    assertThat(p).isNullable();
  }

  @Test
  public void testPartitionNullability2() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.Partition(p1);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testOneOrMoreNullability1() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p = new Patterns.OneOrMore(p1);
    assertThat(p).isNullable();
  }

  @Test
  public void testOneOrMoreNullability2() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.OneOrMore(p1);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testConcurOneOrMoreNullability1() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p = new Patterns.ConcurOneOrMore(p1);
    assertThat(p).isNullable();
  }

  @Test
  public void testConcurOneOrMoreNullability2() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.ConcurOneOrMore(p1);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testRangeIsNotNullable() {
    Pattern p1 = NULLABLE_PATTERN;
    NameClass nameClass = NameClasses.ANY_NAME;
    Pattern p = new Patterns.Range(nameClass, p1);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testEndRangeIsNotNullable() {
    Basics.QName qName = Basics.qName("uri", "localName");
    Pattern p = new Patterns.EndRange(qName, id("id"));
    assertThat(p).isNotNullable();
  }

  @Test
  public void testAfterNullability() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.After(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testAllNullability1() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.All(p1, p2);
    assertThat(p).isNullable();
  }

  @Test
  public void testAllNullability2() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.All(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testAllNullability3() {
    Pattern p1 = NOT_NULLABLE_PATTERN;
    Pattern p2 = NULLABLE_PATTERN;
    Pattern p = new Patterns.All(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testAllNullability4() {
    Pattern p1 = NULLABLE_PATTERN;
    Pattern p2 = NOT_NULLABLE_PATTERN;
    Pattern p = new Patterns.All(p1, p2);
    assertThat(p).isNotNullable();
  }

  @Test
  public void testPatternTreeVisualisationForEmpty() {
    String emptyVisualisation = Utilities.patternTreeToDepth(Patterns.EMPTY, 1);
    assertThat(emptyVisualisation).isEqualTo("Empty()");
  }

  @Test
  public void testPatternTreeVisualisationForNotAllowed() {
    String emptyVisualisation = Utilities.patternTreeToDepth(Patterns.NOT_ALLOWED, 1);
    assertThat(emptyVisualisation).isEqualTo("NotAllowed()");
  }

  @Test
  public void testPatternTreeVisualisationForText() {
    String emptyVisualisation = Utilities.patternTreeToDepth(Patterns.TEXT, 1);
    assertThat(emptyVisualisation).isEqualTo("Text()");
  }

  @Test
  public void testPatternTreeVisualisationForChoice() {
    Pattern choice = choice(text(), empty());
    String emptyVisualisation = Utilities.patternTreeToDepth(choice, 15);
    assertThat(emptyVisualisation).isEqualTo("Choice(\n" +
        "| Text(),\n" +
        "| Empty()\n" +
        ")");
  }

  @Test
  public void testPatternTreeVisualisationForElement() {
    Pattern element = element("book", text());
    String visualisation = Utilities.patternTreeToDepth(element, 10);
    assertThat(visualisation).isEqualTo("Partition(\n" +//
        "| Range(\"book\",\n" +//
        "| | Text()\n" +//
        "| )\n" +//
        ")");
  }
}

File: main/src/test/java/nl/knaw/huygens/alexandria/creole/UtilitiesTest.kt
package nl.knaw.huygens.alexandria.creole

/*
     * #%L
 * alexandria-markup-core
 * =======
 * Copyright (C) 2016 - 2019 HuC DI (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
     */

import nl.knaw.huygens.alexandria.AlexandriaAssertions.assertThat
import nl.knaw.huygens.alexandria.creole.Basics.id
import nl.knaw.huygens.alexandria.creole.patterns.*
import org.junit.Test
import org.slf4j.LoggerFactory


class UtilitiesTest : CreoleTest() {

    @Test
    fun testEmptyIsNullable() {
        val p = Patterns.EMPTY
        assertThat(p).isNullable
    }

    @Test
    fun testNotAllowedIsNotNullable() {
        val p = Patterns.NOT_ALLOWED
        assertThat(p).isNotNullable
    }

    @Test
    fun testTextIsNullable() {
        val p = Patterns.TEXT
        assertThat(p).isNullable
    }

    @Test
    fun testChoiceNullability1() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = Choice(p1, p2)
        assertThat(p).isNullable
    }

    @Test
    fun testChoiceNullability2() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Choice(p1, p2)
        assertThat(p).isNullable
    }

    @Test
    fun testChoiceNullability3() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Choice(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testChoiceNullability4() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = Choice(p1, p2)
        assertThat(p).isNullable
    }

    @Test
    fun testInterleaveNullability1() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = Interleave(p1, p2)
        assertThat(p).isNullable
    }

    @Test
    fun testInterleaveNullability2() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Interleave(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testInterleaveNullability3() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = Interleave(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testInterleaveNullability4() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Interleave(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testConcurNullability1() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = Concur(p1, p2)
        assertThat(p).isNullable
    }

    @Test
    fun testConcurNullability2() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Concur(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testConcurNullability3() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = Concur(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testConcurNullability4() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Concur(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testPartitionNullability1() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p = Partition(p1)
        assertThat(p).isNullable
    }

    @Test
    fun testPartitionNullability2() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = Partition(p1)
        assertThat(p).isNotNullable
    }

    @Test
    fun testOneOrMoreNullability1() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p = OneOrMore(p1)
        assertThat(p).isNullable
    }

    @Test
    fun testOneOrMoreNullability2() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = OneOrMore(p1)
        assertThat(p).isNotNullable
    }

    @Test
    fun testConcurOneOrMoreNullability1() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p = ConcurOneOrMore(p1)
        assertThat(p).isNullable
    }

    @Test
    fun testConcurOneOrMoreNullability2() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = ConcurOneOrMore(p1)
        assertThat(p).isNotNullable
    }

    @Test
    fun testRangeIsNotNullable() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val nameClass = NameClasses.ANY_NAME
        val p = Range(nameClass, p1)
        assertThat(p).isNotNullable
    }

    @Test
    fun testEndRangeIsNotNullable() {
        val qName = Basics.qName("uri", "localName")
        val p = EndRange(qName, id("id"))
        assertThat(p).isNotNullable
    }

    @Test
    fun testAfterNullability() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = After(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testAllNullability1() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = All(p1, p2)
        assertThat(p).isNullable
    }

    @Test
    fun testAllNullability2() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = All(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testAllNullability3() {
        val p1 = CreoleTest.NOT_NULLABLE_PATTERN
        val p2 = CreoleTest.NULLABLE_PATTERN
        val p = All(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testAllNullability4() {
        val p1 = CreoleTest.NULLABLE_PATTERN
        val p2 = CreoleTest.NOT_NULLABLE_PATTERN
        val p = All(p1, p2)
        assertThat(p).isNotNullable
    }

    @Test
    fun testPatternTreeVisualisationForEmpty() {
        val emptyVisualisation = Utilities.patternTreeToDepth(Patterns.EMPTY, 1)
        Assertions.assertThat(emptyVisualisation).isEqualTo("Empty()")
    }

    @Test
    fun testPatternTreeVisualisationForNotAllowed() {
        val emptyVisualisation = Utilities.patternTreeToDepth(Patterns.NOT_ALLOWED, 1)
        Assertions.assertThat(emptyVisualisation).isEqualTo("NotAllowed()")
    }

    @Test
    fun testPatternTreeVisualisationForText() {
        val emptyVisualisation = Utilities.patternTreeToDepth(Patterns.TEXT, 1)
        Assertions.assertThat(emptyVisualisation).isEqualTo("Text()")
    }

    @Test
    fun testPatternTreeVisualisationForChoice() {
        val choice = choice(text(), empty())
        val emptyVisualisation = Utilities.patternTreeToDepth(choice, 15)
        Assertions.assertThat(emptyVisualisation).isEqualTo("Choice(\n" +
                "| Text(),\n" +
                "| Empty()\n" +
                ")")
    }

    @Test
    fun testPatternTreeVisualisationForElement() {
        val element = element("book", text())
        val visualisation = Utilities.patternTreeToDepth(element, 10)
        Assertions.assertThat(visualisation).isEqualTo("Partition(\n" +//

                "| Range(\"book\",\n" +//

                "| | Text()\n" +//

                "| )\n" +//

                ")")
    }

    @Test
    fun testFlip() {
        val p1 = CreoleTest.TestPattern()
        val p2 = CreoleTest.TestPattern()
        val pattern = group(p1, p2)

        val flipped = pattern.flip()
        LOG.info("flipped = ", Utilities.patternTreeToDepth(flipped, 10))
        Assertions.assertThat(flipped.javaClass).isEqualTo(pattern.javaClass)
        val group = pattern as Group
        val flippedGroup = flipped as Group
        assertThat(flippedGroup.pattern1).isEqualTo(group.pattern2)
        assertThat(flippedGroup.pattern2).isEqualTo(group.pattern1)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(UtilitiesTest::class.java!!)
    }

}
